plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -101,34 +105,35 @@ public void failed(Exception e) {
         });
     }
 
-    void processBulkIndexRequest(String action, ActionListener listener, ActionFilterChain chain, BulkRequest bulkRequest, String pipelineId, Iterator<ActionRequest> requests) {
-        if (!requests.hasNext()) {
-            chain.proceed(action, bulkRequest, listener);
+    void processBulkIndexRequest(BulkProcessingState state) {
+        if (!state.hasNext()) {
+            state.proceed();
             return;
         }
 
-        ActionRequest actionRequest = requests.next();
+        ActionRequest actionRequest = state.next();
         if (!(actionRequest instanceof IndexRequest)) {
-            processBulkIndexRequest(action, listener, chain, bulkRequest, pipelineId, requests);
+            processBulkIndexRequest(state);
             return;
         }
 
         IndexRequest indexRequest = (IndexRequest) actionRequest;
         Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
         Data data = new Data(indexRequest.index(), indexRequest.type(), indexRequest.id(), sourceAsMap);
-        executionService.execute(data, pipelineId, new PipelineExecutionService.Listener() {
+        executionService.execute(data, state.pipelineId, new PipelineExecutionService.Listener() {
             @Override
             public void executed(Data data) {
                 if (data.isModified()) {
                     indexRequest.source(data.getDocument());
                 }
-                processBulkIndexRequest(action, listener, chain, bulkRequest, pipelineId, requests);
+                processBulkIndexRequest(state);
             }
 
             @Override
             public void failed(Exception e) {
```

Comment:
wondering if we should change things so we get Throwable here instead of just Exception. is it ok that a Throwable still breaks the whole bulk request?